QuantConnect/Lean
# Refactor OptionExerciseOrder.Quantity to be consistent with other Order types

All order types in LEAN use directional quantities to indicate the change in holdings resulting from filling the order except `OptionExerciseOrder`. This order type was initially set up to only support positive quantities indicating exercise of an options contract. Option assignment was later implemented and used negative quantities. This discrepancy appears to be due to organic evolution and this change aims to bring this order type into alignment with the other order types to maintain consistency of the `Order` abstraction,.

#### Description
* Improve information tracked in regression's {algorithm}.{lang}.details.log

> Greatly improves information tracked for regression tests. See commit message for more details.

* Fix typo in options OrderTests test case name

* Update SymbolRepresentation.GenerationOptionTickerOSI to extension method

> Far more convenient as an extension method

* Improve R# default code formatting rules

> Many of these rule changes focus on improving the readability of code,
with a particular emphasis on multi-line constructs, chained method calls
and multi-line method invocations/declarations.

* Add braces, use string interpolation and limit long lines

* Refactor OptionExerciseOrder.Quantity to indicate change in #contracts

> For all other order types, the Order.Quantity indicates the change in the algorithm's
holdings upon order execution for the order's symbol. For OptionExerciseOrder, this
convention was broken. It appears as though only exercise was initially implemented,
in which case only long positions were supported and a code comment indicated that
only positive values of quantity were acceptable, indicating the number of contracts
to exercise. At a later date, assignment simulation was added and utilized a negative
order quantity. This caused some major inconsistencies in how models view exercise
orders compared to all other order types. This change brings OptionExerciseOrder.Quantity
into alignment with the other order types by making it represent the change in holdings
quantity upon order execution.

> Fixes OptionPortfolioModelTests to go through the exercise model to ensure consistency
and completeness of the test.

> This change was originally going to be much larger, but in order to minimize risks and to
make for an easier review experience, the additional changes will be committed separately
and pushed in their own PR. Some of the issues identified include:
=Manual Exercise (especially for OTM) is not covered
=Margin Calculations (in particular taking into account opposing contracts held)
=Brokerage.OptionPositionAssigned is raised for exercise (later filtered by tx handler)

#### Related Issue




Fixes #4766

#### Motivation and Context

See #4766 TL;DR; `OptionExerciseOrder.Quantity` didn't follow LEAN conventions for order quantity. Upon execution, the order's quantity should represent the change in holdings, but for `OptionExerciseOrder`, positive values indicated an exercise while negative values indicated assignment.

#### Requires Documentation Change

Perhaps. The `QCAlgorithm.ExcerciseOption` method will accept positive or negative values here. Since you must be long the contract, we don't need to validate based on the quantity supplied. The `SubmitOrderRequest.Quantity` ends up being negative to match the `Order.Quantity` as is convention, but for backwards compatibility and also just from the way the method is phrased, it makes sense to accept positive values here. Also, if you look at something like `QCAlgorithm.Buy` or `QCAlgorithm.Sell`, we do something similar, where the direction is implied by the method and we use `Math.Abs(quantity)` to ensure the sign of quantity is correct per the method called.

#### How Has This Been Tested?



Existing unit and regression tests. The original outline of this PR was broader in scope when I wasn't as familiar. As I became more familiar with the stack, it became obvious these quantity consistency changes could be made with a scalpel rather than a hammer.

#### Types of changes

* _kind of_ -- algorithms that depend on `Order.Quantity` or `OrderEvent.Quantity` and are expecting positive to imply exercise and negative to imply assignment would be affected. @jaredbroad -- we need to determine if this is an acceptable level of breaking change. I suspect the actual number of algorithms impacted to be minimal, but worthy of a check.

#### Checklist:
